### PR TITLE
Scenegraph redux

### PR DESCRIPTION
**Pull request by duncanmeech**
Far from working but a much improved version of the scene graph without the React wrappers on all the elements. 

Currently there is a clone of the project page ( dnd ) that you can access with dnd/projectid:id

It creates a ConstructViewer for each construct, which in turns creates a scene graph, layout object etc. Handles basic drag and drop.

Things to note: Work in progress for Friday's demo. Tests coming once there is more to test. Lint is clear but only because I removed the specs folder from linting. Personally I have no problem with slack lint standards in tests. 

Comments please and/or thumbs please.

### CODE REVIEW

**Comment by duncanmeech** [comment]
p.s. I'll rebase to master tomorrow once I see your comments.

File: package.json [hunk 1]
```diff
@@ -7,7 +7,7 @@
     "clean": "rimraf lib dist coverage",
     "check": "npm run lint && npm run test",
     "start": "babel-node devServer.js",
-    "lint": "eslint src test; exit 0;",
+    "lint": "eslint src; exit 0;",
```
Comment (maxbates) [inline package.json]: this is a reversion

File: src/containers/dnd.js [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        </div>
+      </div>
+    );
+  }
+}
+
+function mapStateToProps(state) {
+  const { projectId, constructId } = state.router.params;
+  const project = state.projects[projectId];
+  const constructs = project.components.map(componentId => state.blocks[componentId]);
```
Comment (maxbates) [inline src/containers/dnd.js]: lets strip this?

File: src/containers/graphics/views/constructviewer.js [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    if (type === blockDragType) {
+      // fixme
+      // really, we just need to add it to the store...
+      // going to leave for now... let's discuss how to handle
+      // do we just want an action to associate a block with the store, since this isn't really creating it if its coming from the inventory?
+      // do we want to clone inventory?
+      // What happens when you pull something from inventory to associate with your project?
+      // doing this will use the provided ID, and cause problems in the store
+
+      const block = props.blockCreate(item);
```
Comment (maxbates) [inline src/containers/graphics/views/constructviewer.js]: this wont work with the new promise-y syntax.... also, specifically here, blockCreate should use the new blockClone()

File: src/containers/graphics/views/layout.js [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+ * layout and scene graph manager for the construct viewer
+ */
+export default class Layout {
+
+  constructor(constructViewer, sceneGraph, options) {
+    // we need a construct viewer, a scene graph, a construct and options
+    this.constructViewer = constructViewer;
+    this.sceneGraph = sceneGraph;
+    // set default options
+    Object.assign(this, {
```
Comment (maxbates) [inline src/containers/graphics/views/layout.js]: will need to handle better... now that we are flexbox, this needs to be more accomodating

File: src/containers/graphics/views/layout.js [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+  }
+
+  /**
+   * layout, configured with various options:
+   * xlimit: maximum x extent
+   * @return {[type]} [description]
+   */
+  layout(layoutOptions) {
+    // shortcut
+    const ct = this.construct;
```
Comment (maxbates) [inline src/containers/graphics/views/layout.js]: really?

File: src/containers/graphics/views/layout.js [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+
+      // create element on demand, this should vary by block type, which I don't currently have
+      // this.elementFactory(p, p.type === 'part' ? kT.partAppearance : kT.connectorAppearance);
+      this.elementFactory(part, kT.partAppearance);
+
+      // measure element text or used condensed spacing
+      const td = this.measureText(this.nodeFromElement(part), part, layoutOptions.condensed);
+
+      // if position would exceed x limit then wrap
+      if (xp + td.x > mx) {
```
Comment (maxbates) [inline src/containers/graphics/views/layout.js]: 2 letters isnt helping so much here....

File: src/styles/SceneGraphPage.css [hunk 7]
```diff
@@ -1,16 +1,13 @@
...
-  margin-top: 1rem;
-}
 .sceneGraph {
-  position: absolute;
+  position: relative;
   display: inline-block;
   overflow: hidden;
 }
+/* any element with an animated transform */
+.transform-animated {
```
Comment (maxbates) [inline src/styles/SceneGraphPage.css]: lets do a utils file?

**Comment by maxbates** [comment]
code itself looks more or less fine... would like a run-through on it again.. I dont need to know it through and through but if youre settling on this, would like to know how things work now, how to make new layouts, how to add new elements, etc. etc.

File: package.json (hunk 1 above)
Comment (duncanmeech) [inline package.json]: I don't really want to lint the tests....its annoying and rarely if ever useful.

File: src/containers/dnd.js (hunk 2 above)
Comment (duncanmeech) [inline src/containers/dnd.js]: strip what?

File: src/containers/graphics/views/constructviewer.js (hunk 3 above)
Comment (duncanmeech) [inline src/containers/graphics/views/constructviewer.js]: We can merge when we have a chance to look at your new code together.

File: src/containers/graphics/views/layout.js (hunk 4 above)
Comment (duncanmeech) [inline src/containers/graphics/views/layout.js]: adding responsiveness to size of container as we speak.

File: src/containers/graphics/views/layout.js (hunk 5 above)
Comment (duncanmeech) [inline src/containers/graphics/views/layout.js]: Could you put that in the form a question?

File: src/containers/graphics/views/layout.js (hunk 6 above)
Comment (duncanmeech) [inline src/containers/graphics/views/layout.js]: Sorry, graphics gets messy, lots of boxes, coordinates, radii etc.
Given the liberal sprinkling of comments I added this code is pretty readable.

**Comment by maxbates** [comment]
@duncanmeech is this still valid? Or should be closed?

**Comment by maxbates** [comment]
You'll let me know when this is ready to look at @duncanmeech ?

File: src/containers/graphics/views/layout.js (hunk 5 above)
Comment (maxbates) [inline src/containers/graphics/views/layout.js]: you and your two letter names...

File: src/index.js [hunk 8]
```diff
@@ -14,7 +14,8 @@ render(
   document.getElementById('root')
 );
 
-if (process.env.NODE_ENV !== 'production') {
+//if (process.env.NODE_ENV !== 'production') {
+if (false) {
```
Comment (maxbates) [inline src/index.js]: how about if (`process.env.user === 'maxwellbates'`)

File: src/styles/App.css [hunk 9]
```diff
@@ -3,13 +3,13 @@
 @import 'base.css';
 
 .App {
-  height: 100vh;
   display: flex;
   flex-direction: column;
   background-color: var(--colors-greyLight);
+  min-height: 100vh;
```
Comment (maxbates) [inline src/styles/App.css]: this should stay `height` to prevent scrolling on the page

**Comment by maxbates** [comment]
Notes:

- [ ] use sbol thin / think glyphs, not the switch statement
- [ ] rename the thick symbols to match thin
- [ ] few internal comments

**Pull request merged**